frontend/catalyst/compilation_pipelines.py

```diff
@@ -550,9 +546,102 @@ def __call__(self, *args, **kwargs):
         else:
             args = CompiledFunction.promote_arguments(self.c_sig, r_sig, *args)
 
+        if any(isinstance(arg, jax.core.Tracer) for arg in args):
+            # Only compile a derivative version of the compiled function when needed.
+            if self.jaxed_qfunc is None:
+                self.jaxed_qfunc = JAX_QJIT(self)
+
+            return self.jaxed_qfunc(*args, **kwargs)
+
         return self.compiled_function(*args, **kwargs)
 
 
+class JAX_QJIT:
+    """Wrapper class around :class:`~.QJIT` that enables compatibility with JAX transformations.
+
+    The primary mechanism through which this is effected is by wrapping the invocation of the QJIT
+    object inside a JAX ``pure_callback``. Additionally, a custom JVP is defined in order to support
+    JAX-based differentiation, which is itself a ``pure_callback`` around a second QJIT object which
+    invokes :func:`~.grad` on the original function. Using this class thus incurs additional
+    compilation time.
+
+    Args:
+        qfunc (QJIT): the compiled quantum function object to wrap
+    """
+
+    def __init__(self, qfunc):
+        @jax.custom_jvp
+        def jaxed_qfunc(*args, **kwargs):
+            results = self.wrap_callback(qfunc, *args, **kwargs)
+            if len(results) == 1:
+                results = results[0]
+            return results
+
+        self.qfunc = qfunc
+        self.jaxed_qfunc = jaxed_qfunc
+        self._deriv_qfunc = None  # compiled on demand
+        jaxed_qfunc.defjvp(self.compute_jvp)
+
+    @property
+    def deriv_qfunc(self):
+        """Compile a function computing the derivate of the wrapped QJIT function."""
+        if self._deriv_qfunc is not None:
+            return self._deriv_qfunc
+
+        def deriv_wrapper(*args, **kwargs):
+            # Only consider differentiable arguments (i.e. of type floating point).
+            all_diff_args = [i for i in range(len(args)) if args[i].dtype.kind == "f"]
```

Comment:
Why non-floating args cannot be differentiable?  :thinking:

Reply:
Well typically you don't define derivatives on integer or Boolean-valued functions because the function must be continuous and smooth. You can do complex differentiation but I don't think it's supported in Catalyst right now.

The reason I compute the indices in this way here is to be able to support functions that have differentiable as well as non-differentiable arguments (e.g. `def f(x: float, y: int)`). If I didn't exclude the non-differentiable ones, Catalyst would raise an error at this stage.

Reply:
@josh146 Yes, any tensor with dtype starting with "f".

Reply:
Could users still specify the differentiable arguments? Isn't the assumption of considering all floating points arguments as differentiable args impacting the user experience? Wouldn't it be good to print a warning for cases with non-floating args?

Reply:
> Could users still specify the differentiable arguments? Isn't the assumption of considering all floating points arguments as differentiable args impacting the user experience?

The user specifies argnums at a different level to which we don't have direct access, see my discussion of the issue [under Josh's comment](https://github.com/PennyLaneAI/catalyst/pull/96#discussion_r1173162649). It's mainly a question of performance however, the user experience should not be affected.

> Wouldn't it be good to print a warning for cases with non-floating args?

In the case that a user does somehow define a qjit function on complex numbers, and they also attempt to differentiate it using JAX with holomorphic gradient enabled (which I think this is pretty rare), it would be good to raise an error. I think this means checking whether a tangent vector for which we didn't compute a Jacobian contains non-zero values, which I actually do here:
https://github.com/PennyLaneAI/catalyst/blob/c1f649449c63472dc3deb449f7e5a7abcf386c38/frontend/catalyst/compilation_pipelines.py#L628-L629

Reply:
After implementing the optimization in https://github.com/PennyLaneAI/catalyst/pull/96/commits/f89faf6b71d41069e2f44593b4edf9fd03a36916 this is no longer an issue, we can let the grad op itself raise an error for incorrect types (added here: https://github.com/PennyLaneAI/catalyst/pull/96/commits/efb78c001ae619538d75d86f3590ed3eef3496da, previously those would be raised in MLIR instead).